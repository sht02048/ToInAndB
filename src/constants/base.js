const BASE_PATH = "/assets/sprites/backgrounds/base.png";
const BASE_MODIFIER = {
  GAP: -3,
};

export { BASE_PATH, BASE_MODIFIER };
