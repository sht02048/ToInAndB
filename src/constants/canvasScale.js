const CANVAS_WIDTH = window.innerWidth * 0.6;
const CANVAS_HEIGHT = window.innerHeight;

export { CANVAS_WIDTH, CANVAS_HEIGHT };
