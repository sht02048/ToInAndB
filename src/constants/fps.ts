const FPS = 120;

export default FPS;
