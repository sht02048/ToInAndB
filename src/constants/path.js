const PLAYER_AIRCRAFT_PATH = {
  STATIC: "/assets/sprites/aircraft/player_static.png",
  STRAIGHT: "/assets/sprites/aircraft/player_straight.png",
  LEFT: "/assets/sprites/aircraft/player_left.png",
  RIGHT: "/assets/sprites/aircraft/player_right.png",
};

const BASE_PATH = "/assets/sprites/backgrounds/base.png";

const BLOCK_PATH = {
  LEFT: "/assets/sprites/backgrounds/block_left.png",
  RIGHT: "/assets/sprites/backgrounds/block_right.png",
};

const PLATE_PATH = "/assets/sprites/backgrounds/plate.png";

const SPACE_PATH = "/assets/sprites/backgrounds/space.png";

export { PLAYER_AIRCRAFT_PATH, BASE_PATH, BLOCK_PATH, PLATE_PATH, SPACE_PATH };
