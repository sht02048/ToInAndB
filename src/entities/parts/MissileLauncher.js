import Missile from "./Missile";

class MissileLauncher {
  constructor(width, height) {
    this.x;
    this.y;
    this.width = width;
    this.height = height;

    this.missileList = [];
  }

  load(projectilePath, x, y, missileWidth, missileHeight) {
    const missile = new Missile(projectilePath, missileWidth, missileHeight);

    this.x = x;
    this.y = y;
    missile.x = this.x + this.width / 2 - missile.width / 2;
    missile.y = this.y - 50;

    this.missileList.push(missile);
  }

  render() {
    this.missileList.forEach((missile) => missile.render(missile.x, missile.y));
  }

  update(missileSpeed) {
    this.missileList.forEach((missile) => {
      if (missile.isVanished) {
        return;
      }

      missile.checkVanished();
      missile.launchStraight(missileSpeed);
    });
  }
}

export default MissileLauncher;
