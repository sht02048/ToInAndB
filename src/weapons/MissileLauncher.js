import Missile from "./Missile";

import TEAM from "../constants/team";
import MISSILE_ROUTE_COMMAND from "../constants/missileRouteCommand";

class MissileLauncher {
  #projectileNumber = 15;
  #missileYModifier = 50;
  #enemyDamage = 1;

  constructor(width, height) {
    this.x;
    this.y;
    this.width = width;
    this.height = height;

    this.missileList = [];
  }

  setTargetList(targetList) {
    this.targetList = targetList;
  }

  makeMissile({
    projectilePath,
    x,
    y,
    missileWidth,
    missileDamage = this.#enemyDamage,
    team,
    missileSpeed,
    isAimed = false,
  }) {
    const missile = new Missile(projectilePath);
    this.x = x;
    this.y = y;

    const shipCenter =
      team === TEAM.PLATER
        ? this.x + this.width / 2 - missileWidth / 2
        : this.x + this.width / 2 - missileWidth / 2;

    missile.team = team;
    missile.x = shipCenter;
    missile.y =
      team === TEAM.PLATER
        ? this.y - this.#missileYModifier
        : this.y + this.height - 10;
    missile.damage = missileDamage;
    missile.speed = missileSpeed;
    missile.width = missileWidth;

    if (isAimed) {
      const { vx, vy, angle } = this.getTargetDirection(missile);

      missile.vx = vx;
      missile.vy = vy;
      missile.angle = angle;
    }

    return missile;
  }

  loadSingleAmmo(missileInformation) {
    const missile = this.makeMissile(missileInformation);

    this.missileList.push(missile);
  }

  loadMultipleAmmo(missileInformation) {
    const radiansPerProjectile = (Math.PI * 2) / this.#projectileNumber;

    for (let i = 0; i < this.#projectileNumber; i += 1) {
      const angle = radiansPerProjectile * i;
      const dx = Math.cos(angle);
      const dy = Math.sin(angle);

      const missile = this.makeMissile(missileInformation);

      missile.vx = dx * missile.speed;
      missile.vy = dy * missile.speed;

      this.missileList.push(missile);
    }
  }

  render() {
    this.missileList.forEach((missile) => {
      if (missile.isVanished) {
        return;
      }

      if (missile.angle !== undefined) {
        missile.renderAngle(
          missile.x,
          missile.y,
          missile.width,
          missile.height,
          missile.angle,
        );
        return;
      }

      if (missile.team === TEAM.PLATER) {
        missile.render(missile.x, missile.y);
        return;
      }

      missile.render(missile.x, missile.y);
    });
  }

  setMissileRoute(missileRoute) {
    this.missileList.forEach((missile) => {
      if (missile.isVanished) {
        return;
      }

      switch (missileRoute) {
        case MISSILE_ROUTE_COMMAND.PLATER_STRAIGHT:
          missile.playerStraight(missile.speed);
          break;

        case MISSILE_ROUTE_COMMAND.ENEMY_STRAIGHT:
          missile.enemyStraight(missile.speed);
          break;

        case MISSILE_ROUTE_COMMAND.ENEMY_AIMED:
          missile.enemyTargetMove();
          break;

        case MISSILE_ROUTE_COMMAND.GUIDED:
          const { vx, vy, angle } = this.getTargetDirection(missile);

          if (vx === 0 && vy === 0) {
            if (missile.vx && missile.vy) {
              missile.enemyTargetMove();

              return;
            }

            missile.playerStraight(missile.speed);
            return;
          }

          if (missile.frame < 120) {
            missile.vx = vx;
            missile.vy = vy;
            missile.angle = angle;
          }

          missile.enemyTargetMove();
          break;

        case MISSILE_ROUTE_COMMAND.ENEMY_ALLWAY:
          missile.enemyTargetMove();
          break;
      }

      missile.frame += 1;
    });
  }

  getTargetDirection(missile) {
    if (missile.isLockedOn) {
      return { vx: 0, vy: 0, angle: 0 };
    }

    let minDistance = Infinity;

    const missileVector = {
      vx: 0,
      vy: 0,
      angle: 0,
    };

    this.targetList.forEach((target) => {
      if (target.isDestroyed || target.isVanished) {
        return;
      }

      const targetX = target.x + target.width / 2;
      const targetY = target.y + target.height / 2;
      const missileX = missile.x + missile.width / 2;

      const dx = targetX - missileX;
      const dy = targetY - missile.y;
      const distance = Math.sqrt(dx ** 2 + dy ** 2);
      const normalizedDx = dx / distance;
      const normalizedDy = dy / distance;

      const angle = Math.atan2(dy, dx) + Math.PI / 2;
      const vx = normalizedDx * missile.speed;
      const vy = normalizedDy * missile.speed;

      if (distance < minDistance) {
        minDistance = distance;

        missileVector.vx = vx;
        missileVector.vy = vy;
        missileVector.angle = angle;
      }
    });

    if (missile.isLostTarget) {
      return { vx: 0, vy: 0, angle: 0 };
    }

    missile.isLockedOn = true;

    return missileVector;
  }
}

export default MissileLauncher;
