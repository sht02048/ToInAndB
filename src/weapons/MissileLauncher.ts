import Missile from "./Missile";

import MISSILE_ROUTE_COMMAND from "../constants/missileRouteCommand";
import { MissileInformation } from "../types/interfaces";

import Player from "../entities/Player";
import Enemy from "../entities/enemies/Enemy";
import SpaceShip from "../entities/Spaceship";

class MissileLauncher {
  private target: Player | Enemy;
  private x: number;
  private y: number;
  private width: number;
  private height: number;
  private targetList: SpaceShip[];

  public missileList: Missile[];

  #multipleAngle = 0.1;
  #projectileNumber = 15;
  #enemyDamage = 1;

  constructor(width: number, height: number) {
    this.x;
    this.y;
    this.width = width;
    this.height = height;

    this.missileList = [];
  }

  public setTargetList(targetList: SpaceShip[]): void {
    this.targetList = targetList;
  }

  public makeMissile({
    projectilePath,
    x,
    y,
    missileWidth,
    missileDamage = this.#enemyDamage,
    missileSpeed,
    isAimed = false,
  }: MissileInformation): Missile {
    const missile = new Missile(projectilePath);
    this.x = x;
    this.y = y;

    const shipCenter = this.x + this.width / 2 - missileWidth / 2;

    missile.x = shipCenter;
    missile.y = y;
    missile.damage = missileDamage;
    missile.speed = missileSpeed;
    missile.width = missileWidth;

    if (isAimed) {
      const { vx, vy, angle, targetIndex } = this.#lockOn(missile);

      missile.vx = vx;
      missile.vy = vy;
      missile.angle = angle;
      missile.targetIndex = targetIndex;
    }

    return missile;
  }

  public loadSingleMissile(missileInformation: MissileInformation): void {
    const missile = this.makeMissile(missileInformation);

    this.missileList.push(missile);
  }

  public loadMultipleMissile(missileInformation: MissileInformation): void {
    const radiansPerProjectile = (Math.PI * 2) / this.#projectileNumber;

    for (let i = 0; i < this.#projectileNumber; i += 1) {
      const angle = radiansPerProjectile * i + this.#multipleAngle;
      const dx = Math.cos(angle);
      const dy = Math.sin(angle);

      const missile = this.makeMissile(missileInformation);

      missile.vx = dx * missile.speed;
      missile.vy = dy * missile.speed;

      this.missileList.push(missile);
    }

    if (missileInformation.shouldTilt) {
      this.#multipleAngle += 0.02;
    }
  }

  public render(): void {
    this.missileList.forEach((missile: Missile): void => {
      if (missile.isVanished) {
        return;
      }

      if (missile.angle !== undefined) {
        missile.renderAngle(
          missile.x,
          missile.y,
          missile.width,
          missile.height,
          missile.angle,
        );
        return;
      }

      missile.render(missile.x, missile.y);
    });
  }

  setMissileRoute(missileRoute: string): void {
    this.missileList.forEach((missile: Missile): void => {
      if (missile.isVanished) {
        return;
      }

      switch (missileRoute) {
        case MISSILE_ROUTE_COMMAND.PLAYER_STRAIGHT: {
          missile.playerStraight(missile.speed);
          break;
        }

        case MISSILE_ROUTE_COMMAND.ENEMY_STRAIGHT: {
          missile.enemyStraight(missile.speed);
          break;
        }

        case MISSILE_ROUTE_COMMAND.ENEMY_AIMED: {
          missile.enemyTargetMove();
          break;
        }

        case MISSILE_ROUTE_COMMAND.GUIDED: {
          const { vx, vy, angle, targetIndex } = this.#lockOn(missile);

          missile.targetIndex = targetIndex;

          if (vx === 0 && vy === 0) {
            if (missile.vx && missile.vy) {
              missile.enemyTargetMove();

              return;
            }

            missile.playerStraight(missile.speed);
            return;
          }

          if (missile.frame < 160) {
            missile.vx = vx;
            missile.vy = vy;
            missile.angle = angle;
          }

          missile.enemyTargetMove();
          break;
        }

        case MISSILE_ROUTE_COMMAND.ENEMY_ALLWAY: {
          missile.enemyTargetMove();
          break;
        }
      }

      missile.frame += 1;
    });
  }

  #lockOn(missile: Missile) {
    const missileVector = {
      vx: 0,
      vy: 0,
      angle: 0,
      targetIndex: null,
    };

    if (missile.targetIndex !== null) {
      const target = this.targetList[missile.targetIndex];
      missileVector.targetIndex = missile.targetIndex;

      if (target === undefined || target.isDestroyed || target.isVanished) {
        return missileVector;
      }

      return this.#getTargetDirection(target, missile);
    }

    let minDistance = Infinity;

    this.targetList.forEach((target, index) => {
      if (target.isDestroyed || target.isVanished) {
        return;
      }

      const { vx, vy, angle, distance } = this.#getTargetDirection(
        target,
        missile,
      );

      if (distance < minDistance) {
        minDistance = distance;

        missileVector.vx = vx;
        missileVector.vy = vy;
        missileVector.angle = angle;
        missileVector.targetIndex = index;
      }
    });

    return missileVector;
  }

  #getTargetDirection(target: SpaceShip, missile: Missile) {
    const targetX = target.x + target.width / 2;
    const targetY = target.y + target.height / 2;
    const missileX = missile.x + missile.width / 2;

    const dx = targetX - missileX;
    const dy = targetY - missile.y;
    const distance = Math.sqrt(dx ** 2 + dy ** 2);
    const normalizedDx = dx / distance;
    const normalizedDy = dy / distance;

    const angle = Math.atan2(dy, dx) + Math.PI / 2;
    const vx = normalizedDx * missile.speed;
    const vy = normalizedDy * missile.speed;

    return {
      vx,
      vy,
      angle,
      targetIndex: missile.targetIndex,
      distance,
    };
  }
}

export default MissileLauncher;
