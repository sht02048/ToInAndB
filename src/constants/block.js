import { BLOCKS } from "./path";

const LEFT_BLOCKS = [BLOCKS.FIRST_LEFT, BLOCKS.SECOND_LEFT, BLOCKS.THIRD_LEFT];

const RIGHT_BLOCKS = [
  BLOCKS.FIRST_RIGHT,
  BLOCKS.SECOND_RIGHT,
  BLOCKS.THIRD_RIGHT,
];

const BLOCK_LIST = [
  {
    GROUP: "left",
    ITEMS: [
      { img: new Image(), src: BLOCKS.FIRST_LEFT },
      { img: new Image(), src: BLOCKS.SECOND_LEFT },
      { img: new Image(), src: BLOCKS.THIRD_LEFT },
    ],
  },
  {
    GROUP: "right",
    ITEMS: [
      { img: new Image(), src: BLOCKS.FIRST_RIGHT },
      { img: new Image(), src: BLOCKS.SECOND_RIGHT },
      { img: new Image(), src: BLOCKS.THIRD_RIGHT },
    ],
  },
];

const BLOCK_SCALE = 2;
const BLOCK_ID = "block-canvas";

export { BLOCK_LIST, BLOCK_SCALE, BLOCK_ID, LEFT_BLOCKS, RIGHT_BLOCKS };
