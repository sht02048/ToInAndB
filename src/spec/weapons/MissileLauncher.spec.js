import MissileLauncher from "../../weapons/MissileLauncher";

import { PROJECTILE } from "../../constants/path";
import MISSILE_ROUTE_COMMAND from "../../constants/missileRouteCommand";

describe("missileLauncher", () => {
  const mainCanvas = document.createElement("canvas");
  const introCanvas = document.createElement("canvas");
  mainCanvas.id = "main-canvas";
  introCanvas.id = "intro-canvas";

  document.body.appendChild(mainCanvas);
  document.body.appendChild(introCanvas);

  const missileY = 10;
  const missileX = 10;

  const singleMissileInformation = {
    projectilePath: PROJECTILE.LEVEL_1,
    x: missileX,
    y: missileY,
    missileWidth: 10,
    missileSpeed: 1,
    isAimed: true,
  };
  const multipleMissileInformation = {
    projectilePath: PROJECTILE.LEVEL_1,
    x: missileX,
    y: missileY,
    missileWidth: 10,
    missileSpeed: 1,
    isAimed: false,
    shouldTilt: true,
  };

  let missileLauncher;
  let missile;
  let target;
  let missileList;

  beforeEach(() => {
    target = {
      x: 0,
      y: 0,
      width: 10,
      height: 10,
    };

    missileLauncher = new MissileLauncher(32, 32);
    missileLauncher.setTargetList([target]);
    missileList = missileLauncher.missileList;
  });

  it("should have proper functions", () => {
    expect(missileLauncher.setTargetList).to.be.a("function");
    expect(missileLauncher.makeMissile).to.be.a("function");
    expect(missileLauncher.loadSingleMissile).to.be.a("function");
    expect(missileLauncher.loadMultipleMissile).to.be.a("function");
    expect(missileLauncher.render).to.be.a("function");
    expect(missileLauncher.setMissileRoute).to.be.a("function");
  });

  it("should load single missile", () => {
    missileLauncher.loadSingleMissile(singleMissileInformation);
    missile = missileList[0];

    expect(missile.y).to.equal(missileY);
    expect(missile.width).to.equal(10);
    expect(missile.speed).to.equal(1);
    expect(missile.vx).to.be.a("number");
    expect(missile.vy).to.be.a("number");
    expect(missile.angle).to.be.a("number");
    expect(missile.targetIndex).to.equal(0);
  });

  it("should load multiple missiles", () => {
    missileLauncher.loadMultipleMissile(multipleMissileInformation);

    expect(missileList.length).to.equal(15);
    expect(missileList.length).to.equal(15);
  });

  it("should render missile", () => {
    missileLauncher.loadSingleMissile(singleMissileInformation);
    missileLauncher.render();

    const ctx = mainCanvas.getContext("2d");
    const events = ctx.__getEvents();

    expect(events).toMatchSnapshot();
  });

  describe("should setMissileRoute of", () => {
    let shipCenter;
    beforeEach(() => {
      missileLauncher.loadSingleMissile(singleMissileInformation);
      missile = missileList[0];
      shipCenter = missile.x;
    });

    it("player straight", () => {
      missileLauncher.setMissileRoute(MISSILE_ROUTE_COMMAND.PLAYER_STRAIGHT);

      expect(missile.y).to.equal(missileY - missile.speed);
    });

    it("enemy straight", () => {
      missileLauncher.setMissileRoute(MISSILE_ROUTE_COMMAND.ENEMY_STRAIGHT);

      expect(missile.y).to.equal(missileY + missile.speed);
    });

    it("enemy aimed", () => {
      missileLauncher.setMissileRoute(MISSILE_ROUTE_COMMAND.ENEMY_AIMED);

      expect(missile.x).to.equal(shipCenter + missile.vx);
      expect(missile.y).to.equal(missileY + missile.vy);
    });

    it("enemy guided", () => {
      missileLauncher.setMissileRoute(MISSILE_ROUTE_COMMAND.GUIDED);

      expect(missile.x).to.equal(shipCenter + missile.vx);
      expect(missile.y).to.equal(missileY + missile.vy);
    });

    it("enemy allway", () => {
      missileLauncher.setMissileRoute(MISSILE_ROUTE_COMMAND.ENEMY_ALLWAY);

      expect(missile.x).to.equal(shipCenter + missile.vx);
      expect(missile.y).to.equal(missileY + missile.vy);
    });
  });
});
