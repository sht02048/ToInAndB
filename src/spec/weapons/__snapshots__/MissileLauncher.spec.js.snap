// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`missileLauncher > should render missile 1`] = `
[
  {
    "props": {},
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "save",
  },
  {
    "props": {
      "x": 21,
      "y": 20,
    },
    "transform": [
      1,
      0,
      0,
      1,
      21,
      20,
    ],
    "type": "translate",
  },
  {
    "props": {
      "angle": -1.3370531459259953,
    },
    "transform": [
      0.2316205273060396,
      -0.9728062146853669,
      0.9728062146853669,
      0.2316205273060396,
      21,
      20,
    ],
    "type": "rotate",
  },
  {
    "props": {},
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "restore",
  },
]
`;
